static/js/components/dashboard/CourseListCard.js

```diff
@@ -11,54 +11,196 @@ import type {
 } from '../../flow/couponTypes';
 import CourseRow from './CourseRow';
 import FinancialAidCalculator from '../../containers/FinancialAidCalculator';
+import type { CoursePrice } from '../../flow/dashboardTypes';
+import type { CourseRun } from '../../flow/programTypes';
+import {
+  FA_TERMINAL_STATUSES,
+  COUPON_CONTENT_TYPE_PROGRAM,
+  COUPON_AMOUNT_TYPE_PERCENT_DISCOUNT,
+  COUPON_AMOUNT_TYPE_FIXED_DISCOUNT,
+} from '../../constants';
+import {
+  makeAmountMessage,
+  makeCouponReason,
+  isFreeCoupon,
+} from '../../lib/coupon';
+import { pickExistingProps } from '../../util/util';
+
+const priceMessageClassName = "price-message";
 
 export default class CourseListCard extends React.Component {
   props: {
-    coupon:                       ?Coupon,
+    coupon?:                      Coupon,
     program:                      Program,
-    courseEnrollAddStatus?:       string,
+    coursePrice:                  CoursePrice,
     prices:                       CalculatedPrices,
     openFinancialAidCalculator?:  () => void,
     now?:                         Object,
-    addCourseEnrollment:          (courseId: string) => void,
-    openCourseContactDialog:      (course: Course, canContactCourseTeam: boolean) => void
+    addCourseEnrollment:          (courseId: string) => Promise<*>,
+    openCourseContactDialog:      (course: Course, canContactCourseTeam: boolean) => void,
+    setEnrollSelectedCourseRun:   (r: CourseRun) => void,
+    setEnrollCourseDialogVisibility: (bool: boolean) => void,
   };
 
-  render() {
-    let {
-      program,
+  sortedCourseRuns(): Array<CourseRun> {
+    const { program } = this.props;
+    const sortedCourses = R.sortBy(R.prop('position_in_program'), program.courses);
+    return R.unnest(
+      R.map(R.sortBy(R.prop('position')), R.pluck('runs', sortedCourses))
+    );
+  }
+
+  renderFinancialAidPriceMessage(): React$Element<*> {
+    const {
+      program: {financial_aid_user_info: {application_status: finAidStatus}},
+      coupon,
       prices,
-      now,
+    } = this.props;
+
+    if (FA_TERMINAL_STATUSES.includes(finAidStatus)) {
+      let price;
+      for (const courseRun of this.sortedCourseRuns()) {
+        price = prices.get(courseRun.id);
+        if (price) {
+          break;
+        }
+      }
+
+      if (!price) {
+        throw "missing price"; // this should never happen, it's just to make flow happy
```

Comment:
Could that ever actually happen? I'm throwing an error rather than returning null, so that I can define the flow type of this function to _always_ return a React element. If it's possible to return a null, then that's fine, but I'll need to change the flow type.